url85940d3d8e_ccp999_Expr4j_tgz-pJ8-org_boris_expr_util_ExprEvaluatorJ8/src/org/boris/expr/parser/ExprParser.java

```diff
@@ -0,0 +1,322 @@
+        return p.get();
+    }
+
+    public void setParserVisitor(IParserVisitor visitor) {
+        this.visitor = visitor;
+    }
+
+    public void parse(ExprLexer lexer) throws IOException, ExprException {
+        ExprToken e = null;
+        while ((e = lexer.next()) != null) {
+            parseToken(lexer, e);
+        }
+    }
+
+    private void parseToken(ExprLexer lexer, ExprToken token)
+            throws ExprException, IOException {
+        switch (token.type) {
+        case Plus:
+        case Minus:
+        case Multiply:
+        case Divide:
+        case Power:
+        case StringConcat:
+        case LessThan:
+        case LessThanOrEqualTo:
+        case GreaterThan:
+        case GreaterThanOrEqualTo:
+        case Equal:
+        case NotEqual:
+            parseOperator(token);
+            break;
+        case Decimal:
+        case Integer:
+        case String:
+        case Variable:
+            parseValue(token);
+            break;
+        case OpenBracket:
+            parseExpression(lexer);
+            break;
+        case Function:
+            parseFunction(token, lexer);
+            break;
+        case OpenBrace:
+            parseArray(lexer);
+            break;
+        default:
+            throw new ExprException("Unexpected " + token.type + " found");
+        }
+    }
+
+    private void parseFunction(ExprToken token, ExprLexer lexer)
+            throws ExprException, IOException {
+        Expr c = current;
+        current = null;
+        ExprToken e = null;
+        ArrayList args = new ArrayList();
+        while ((e = lexer.next()) != null) {
+            if (e.type.equals(ExprTokenType.Comma)) {
+                if (current == null)
+                    args.add(ExprMissing.MISSING);
+                else
+                    args.add(current);
+                current = null;
+            } else if (e.type.equals(ExprTokenType.CloseBracket)) {
+                if (current != null)
+                    args.add(current);
+                current = c;
+                break;
+            } else {
+                parseToken(lexer, e);
+            }
+        }
+
+        ExprFunction f = new ExprFunction(token.val, (Expr[]) args
+                .toArray(new Expr[0]));
+
+        if (visitor != null)
+            visitor.annotateFunction(f);
+
+        setValue(f);
+    }
+
+    private void parseExpression(ExprLexer lexer) throws IOException,
+            ExprException {
+        Expr c = current;
+        current = null;
+        ExprToken e = null;
+        while ((e = lexer.next()) != null) {
+            if (e.type.equals(ExprTokenType.CloseBracket)) {
+                Expr t = current;
+                current = c;
+                setValue(new ExprExpression(t));
+                break;
+            } else {
+                parseToken(lexer, e);
+            }
+        }
+    }
+
+    private void parseArray(ExprLexer lexer) throws ExprException, IOException {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 13. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.